File: sentimental-credit/credit.py
from sys import exit

while True:
    number = input("Number: ")

    # Try converting height to int
    try:
        int(number)
    except:
        # If exception, go back to start of while loop
        continue
    break

length = len(number)
total = 0
for i in range(len(number) - 1, 0, -2):
    # Loop over every other digit startingfrom the first
    num = int(number[i - 1])
    num *= 2
    # Add individual digits
    if num >= 10:
        n1 = int(num % 10)
        n2 = int(num % 100 / 10)
        total += n1
        total += n2
    else:
        total += num

# Add sum to the numbers that weren't multiplied by 2
for i in range(len(number) - 1, -1v, -2):
    total += int(number[i])

# Validate Luhn's algorithm
if total % 10 != 0:
    print("INVALID")
    exit()

# Validate American Express
if length == 15:
    if number[0] == '3':
        if number[1] == '4' or number[1] == '7':
            print("AMEX")
            exit()

# Validate MasterCard
if length == 16:
    if number[0] == '5':
        if int(number[1]) >= 1 and int(number[1]) <= 5:
            print("MASTERCARD")
            exit()

# Validate Visa
if length == 13 or length == 16:
    if number[0] == '4':
        print("VISA")
        exit()

# If card matches none of the Credit number patterns, that means it is invalid
print("INVALID")

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    long credit = get_long("Number: ");
    int length = floor(log10(credit)) + 1;
    long num = 0;
    long modulo = 0;
    long divider = 0;
    long sum = 0;
    int first = 0;
    int second = 0;

    if (length >= 13 && length <= 16)
    {
        for (int i = 2; i <= length; i += 2)
        {
            modulo = pow(10, i);
            divider = pow(10, i - 1);
            num = credit % modulo / divider;
            num *= 2;
            if (num >= 10)
            {
                second = num % 10;
                first = (num % 100) / 10;
                sum += first;
                sum += second;
            }
            else
            {
                sum += num;
            }
        }

        for (int i = 1; i <= length; i += 2)
        {
            modulo = pow(10, i);
            divider = pow(10, i - 1);
            num = credit % modulo / divider;
            sum += num;
        }

        if (sum % 10 == 0)
        {

            // Debit is valid: Determine which type of debit card it is

            // AMERICAN EXPRESS - Length 15 - starts with 34 or 37
            if (length == 15)
            {
                long div = pow(10, 13);
                long starting_number = credit / div;
                if (starting_number == 34 || starting_number == 37)
                {
                    printf("AMEX\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }
            // VISA - Length 13 - starts with 4
            else if (length == 13)
            {
                long div = pow(10, 12);
                long starting_number = credit / div;
                if (starting_number == 4)
                {
                    printf("VISA\n");
                }
            }

            // VISA - Length 16 - starts with 4
            // MASTERCARD - Length 16 - starts with 51, 52, 53, 54, or 55
            else if (length == 16)
            {
                long starting_number = credit / pow(10, 15);
                if (starting_number == 4)
                {
                    printf("VISA\n");
                }
                else
                {
                    starting_number = credit / pow(10, 14);
                    if (starting_number >= 51 && starting_number <= 55)
                    {
                        printf("MASTERCARD\n");
                    }
                    else
                    {
                        printf("INVALID\n");
                    }
                }
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}